### ISSUE TITLE
Big variations at `ConfusionMatrix` and `featureImportances`

### ISSUE THREAD

**Issue by batata004** [RubixML/ML#98]
Hi,

Why does running the code below many times I get so many variations to `ConfusionMatrix` and `featureImportances`? Mainly to `featureImportances`, the fluctuations are so huge! Sometimes a feature is listed as more important, later, the same feature, is listed as least important.

My code is below (make sure you save as UTF-8). The `file.csv` can be downloaded from [https://www.quemfazsite.com.br/temp/file.csv](https://www.quemfazsite.com.br/temp/file.csv)

Note: I scrumbled the csv file contents just to make sure I dont infringe any privacy policy from the company I work for, also I reduced the dataset to just over 1k lines to make sure I am not providing too much information (despite it all being scrambled, I dont want my boss complaning to me that I wouldnt need to share the entire csv).

```php
<?php



//Este arquivo deve ser salvo com codificação UTF-8.



set_time_limit(0); 
ini_set("memory_limit","-1");
ini_set("pcre.backtrack_limit",1000000000);



echo '<meta charset="UTF-8">';



include __DIR__ . '/vendor/autoload.php';

use Rubix\ML\Datasets\Labeled;
use Rubix\ML\Extractors\CSV;
use Rubix\ML\Extractors\ColumnPicker;
use Rubix\ML\CrossValidation\Reports\ConfusionMatrix;
use Rubix\ML\Classifiers\RandomForest;
use Rubix\ML\Classifiers\ClassificationTree;
use Rubix\ML\Transformers\OneHotEncoder;



$coluna = array("Dispositivo","Dia da semana","Sexo","Idade","Estado","Hora do dia","Conversões");
$dataset = Labeled::fromIterator(new ColumnPicker(new CSV("file.csv",true),$coluna));

$transformer = new OneHotEncoder();
$dataset->apply($transformer);

$temp43 = array();
for ($i=0;$i<count($transformer->categories());$i++) {
	
	for ($j=0;$j<count($transformer->categories()[$i]);$j++) {

		$temp43[] = $coluna[$i] . ";" . $transformer->categories()[$i][$j];
		
	}
	
}

$coluna = $temp43;







//This code is necessary because "OneHotEncoder" returns all columns as a number, so unfortunatelly I have to convert them to strings to be considered as cattegories by the model (as it only accepts cattegories).
for ($i=0;$i<count($dataset->samples()[0]);$i++) {
	
	$dataset->transformColumn(

		$i,
		function ($value){
				
			if (   ($value === "0")   or   ($value === 0)   or   ($value === "Não")   ) {
				
				return "N";
				
			}
			else if (   ($value === "1")   or   ($value === 1)   or   ($value === "Sim")   ) {
				
				return "S";
				
			}
			else{
				
				echo 'Erro PW23. Valor: "' . $value . '".';
				exit();
				//return $value;
				
			}
			
		}
		
	);

}

$dataset->transformLabels(

	function ($value){

		if (   ($value === "0")   or   ($value === 0)   or   ($value === "Não")   ) {
			
			return "N";
			
		}
		else if (   ($value === "1")   or   ($value === 1)   or   ($value === "Sim")   ) {
			
			return "S";
			
		}
		else{
			
			echo 'Erro JH54. Valor: "' . $value . '".';
			exit();
			//return $value;
			
		}
		
	}
	
);





$training = $dataset;
$testing = $dataset;

$estimator = new RandomForest(new ClassificationTree(10),300,0.1,true);	

$estimator->train($training);






$predictions = $estimator->predict($testing);








$importances = $estimator->featureImportances();

asort($importances);

foreach ($importances as $chave1 => $valor1) {

	echo $coluna[$chave1] . " -> " . $valor1 . "\n";

}








$report = new ConfusionMatrix();
$result = $report->generate($predictions,$testing->labels());

var_dump($result);







?>
```

**Comment by andrewdalpino** [comment]
I see a number of things not quite right with this script but I'll just focus on the stuff that relates to the results of your confusion matrix and feature importances ...

First and foremost, you should not use any training data to test your model. In your case the training set is the same data as the testing set, so this is as far as you can get from measuring generalization performance of the model via the confusion matrix. Please see the docs regarding cross validation here ...

https://docs.rubixml.com/en/latest/cross-validation.html

Regarding feature importances ... feature importances measure the degree to which a particular feature contributes to (or is represented in) the model. It computes these values from the training data. Fluctuations in the model will naturally cause changes in the importances of each feature since Random Forest is a non-deterministic algorithm.

https://docs.rubixml.com/en/latest/classifiers/random-forest.html

https://docs.rubixml.com/en/latest/training.html#feature-importances

Problems like overfitting and underfitting can also cause high variance in feature importances, but we would not be able to determine this with your script because of the first issue.

**Comment by batata004** [comment]
Thank you again @andrewdalpino . Now I am at a meeting but just a guy is talking so I am trying to reply you.

First, with the code above I tried my best to remove any function that might change too much the data. That's why I simply used the same dataset as traning and testing. But in my original code, I use `[$training, $testing] = $dataset->stratifiedSplit(0.8);` and the problem of LOTS of fluctuations remains.

So, in the code above please replace:

```php
$training = $dataset;
$testing = $dataset;
```

With just:

`[$training, $testing] = $dataset->stratifiedSplit(0.8);` 

And even doing so, you will see the same big fluctuations in the feature importances and some fluctuations on confusion matrix.

See, I am not saying my dataset is perfect, or that I am trying to build a model using the best conditions to get the best accuracy possible. No, at this point I am just trying to get some stable results from Rubix ML models. So far I have been having mixed results using Rubix ML, and I am at a point where I am trying really hard to first come up with a stable model, then, when I start trusting the model and the library, I will come back and try to optimize my model.

Just so you know, my original code was much longer than the one I posted above. I tried a lot of different things already, A LOT. And after lots of frustration I finally pined point that the problems I am facing are first because of lack of stability of the model, everytime I get completely different results. I will just paste you my original code so, if you have some time, you can see lots of things that I already tried.

I heavily commented all of my code so I am always confident of what I am doing (I know everything is in portuguese and you might not understand my comments, but you can see my big efforts trying to make this Rubix ML work!).



```php
<?php



//Este arquivo deve ser salvo com codificação UTF-8.



set_time_limit(0); 
ini_set("memory_limit","-1");
ini_set("pcre.backtrack_limit",1000000000);



echo '<meta charset="UTF-8">';



include __DIR__ . '/vendor/autoload.php';

use Rubix\ML\Datasets\Labeled;
use Rubix\ML\Extractors\JSON;
use Rubix\ML\Extractors\CSV;
use Rubix\ML\Classifiers\KNearestNeighbors;
use Rubix\ML\CrossValidation\Metrics\Accuracy;
use Rubix\ML\Persisters\Filesystem;
use Rubix\ML\PersistentModel;
use Rubix\ML\Transformers\NumericStringConverter;
use Rubix\ML\Extractors\ColumnPicker;
use Rubix\ML\CrossValidation\Reports\ConfusionMatrix;
use Rubix\ML\CrossValidation\Metrics\FBeta;
use Rubix\ML\Kernels\Distance\Hamming;
use Rubix\ML\Classifiers\NaiveBayes;
use Rubix\ML\Other\Loggers\Screen;
use Rubix\ML\Classifiers\RandomForest;
use Rubix\ML\Classifiers\ClassificationTree;
use Rubix\ML\Transformers\OneHotEncoder;
use Rubix\ML\Transformers\RecursiveFeatureEliminator;
use Rubix\ML\Regressors\RegressionTree;



echo "Carregando informações...\n";

//Se utilizar "new CSV" na linha de código abaixo (com o segundo parâmetro sendo "true", indicando que a primeira linha do arquivo CSV é o cabeçalho) então pode-se filtrar somente as colunas que se deseja utilizar com "ColumnPicker" através do nome delas, por exemplo, ao invés de -->>[0,1,2]<<-- poder-se-ia utilizar -->>["COMPUTADOR","CELULAR"]<<--. Além disso, "ColumnPicker" reordena as colunas no dataset então pode-se utilizar essa função para escolher qual coluna deve ser a última, independentemente da ordem que aparecem no arquivo CSV.
//$dataset = Labeled::fromIterator(new ColumnPicker(new JSON("teste1.json"),[0,1,2]));
//"Dispositivo","Dia da semana","Sexo","Idade","Estado","Hora do dia","Conversões"
//"Fechou negócio" ou "Conversões"
$coluna = array("Dispositivo","Dia da semana","Sexo","Idade","Estado","Hora do dia","Conversões");
$dataset = Labeled::fromIterator(new ColumnPicker(new CSV("relatorio-6.csv",true),$coluna));


//A linha de código abaixo só deve ser utilizada quando se utilizar "new CSV" e se o modelo a ser usado exigir samples numéricos pois quando se utiliza "new CSV" todo o dataset é importado como string e para que alguns modelos funcionem, como "KNearestNeighbors", todas as colunas, exceto a última (output), devem ser numéricas. Além disso, em alguns modelos será necessário comentar a linha de código abaixo pois podem não aceitar valores numéricos nas colunas, então os valores das colunas devem ser strings.
//$dataset->apply(new NumericStringConverter());



//A linha de código abaixo percorre todas as colunas de valor categórico, para cada valor distinto dentro dessa coluna é criada uma nova coluna e após isso o algoritmo preenche cada coluna com 0 ou 1. Esse código substitui por completo a função do arquivo "preparar_csv.php".
$transformer = new OneHotEncoder();
$dataset->apply($transformer);

$temp43 = array();
for ($i=0;$i<count($transformer->categories());$i++) {
	
	for ($j=0;$j<count($transformer->categories()[$i]);$j++) {

		$temp43[] = $coluna[$i] . ";" . $transformer->categories()[$i][$j];
		
	}
	
}

$coluna = $temp43;






//O comando abaixo apenas exibirá algumas poucas colunas (samples) e o label para algumas poucas linhas.
echo $dataset;






//Remove as linhas duplicadas considerando o valor de todas as colunas, tanto nas colunas "samples" e "labels".
//$dataset->deduplicate();

//Percorre todos os elementos dentro de "samples", na coluna especificada no primeiro parâmetro, e se pode realizar alterações. Como "OneHotEncoder" atribui valor numérico 0 ou 1 para cada coluna, então deve-se converter esse valor numérico para um valor categórico pois alguns modelos exigem valor categórico. Alé disso não é desejável que alguns modelos interpretem o valor das colunas como numérico mas sim como categórico pois o número 0 ou 1 não significa nada.
for ($i=0;$i<count($dataset->samples()[0]);$i++) {
	
	$dataset->transformColumn(

		$i,
		function ($value){
				
			if (   ($value === "0")   or   ($value === 0)   or   ($value === "Não")   ) {
				
				return "N";
				
			}
			else if (   ($value === "1")   or   ($value === 1)   or   ($value === "Sim")   ) {
				
				return "S";
				
			}
			else{
				
				echo 'Erro PW23. Valor: "' . $value . '".';
				exit();
				//return $value;
				
			}
			
		}
		
	);

}

//Percorre todos os elementos dentro de "labels" e se pode realizar alterações. Por conta de um "bug" no próprio PHP, nenhum item dentro dos "labels" pode conter um valor numérico, mesmo que esse valor numérico esteja representado como uma string pois o PHP, por razões estranhas e especificamente quando se usa a biblioteca Rubix ML, converterá as strings numéricas para tipo numérico e alguns modelos de ML não permitem tipo numérico dentro dos "labels". Então, nesses casos específicos, deve-se converter os "labels" para algum valor que contenha letras.
$dataset->transformLabels(

	function ($value){

		if (   ($value === "0")   or   ($value === 0)   or   ($value === "Não")   ) {
			
			return "N";
			
		}
		else if (   ($value === "1")   or   ($value === 1)   or   ($value === "Sim")   ) {
			
			return "S";
			
		}
		else{
			
			echo 'Erro JH54. Valor: "' . $value . '".';
			exit();
			//return $value;
			
		}
		
	}
	
);



$nome_arquivo_persistencia_dados_modelo = "modelo.model";

if (isset($_GET["testar"])) {
		
	echo "Carregando dados de treino salvos anteriormente...\n";
	
	$persister = new Filesystem($nome_arquivo_persistencia_dados_modelo);
	$estimator = $persister->load();
	
	$testing = $dataset;
	
}
else{

	echo "Dividindo aleatoriamente os " . count($dataset->labels()) . " elementos para treino e teste...\n";

	$dataset->randomize();

	[$training, $testing] = $dataset->stratifiedSplit(0.8);

	echo "Restaram " . count($training->labels()) . " elemento(s) para treino e " . count($testing->labels()) . " elemento(s) para teste...\n";



	echo "Definindo modelo de treino...\n";
	
	//O kernel padrão do modelo "KNearestNeighbors" é "Euclidean" que, supostamente, funciona bem com valores contínuos nas colunas. Alterando-se o kernel para "Hamming" é mais apropriado pois ele, supostamente, funcione bem com valores de categoria nas colunas.
	//$kernel = new Hamming();
	//$estimator = new KNearestNeighbors(3,true,$kernel);
	
	//$estimator = new NaiveBayes(1);	
		
	$estimator = new RandomForest(new ClassificationTree(10),300,0.1,true);	
		
	echo "Treinando...\n";


	//Alguns modelos permitem acompanhar a evolução das iterações.
	//$estimator->setLogger(new Screen());

	$estimator->train($training);

	$persister = new Filesystem($nome_arquivo_persistencia_dados_modelo);
	$persister->save($estimator);
	
}



echo "Fazendo previsões em cima dos elementos de teste...\n";

//var_dump($testing);exit();

$predictions = $estimator->predict($testing);


//print_r(array_slice($predictions,0,500));
for ($i=0;$i<min(count($testing->labels()),500);$i++) {

	echo $i . " ->" . implode(",",$testing->samples()[$i]) . " -> " . $testing->labels()[$i] . " -> " . $predictions[$i] . " -> " . ($testing->labels()[$i] == $predictions[$i]?"OK":"NOT OK") . "\n";
	
}


//No caso particular que se está analisando, é muito mais importante que a métrica de análise penalize mais duramente os casos em que o usuário de fato iria converter (actual) porém o algoritmo preveu que não iria converter do que os casos em que o usuário não iria converter (actual) porém o algoritmo preveu que iria converter. Ou seja, o "falso negativo" é muito pior que o "falso positivo". Nesse caso, o ideal é utilizar "FBeta" ao invés de "Accuracy".
//$metric = new Accuracy();
//Quanto mais baixo o parâmetro da função abaixo, mais se favorece o "precision" e quanto mais alto o parâmetro da função abaixo, mais se favorece o "recall".
$metric = new FBeta(1);

$score = $metric->score($predictions,$testing->labels());

echo "A precisão do modelo, com base nos elementos de teste, é de " . ($score*100) . "%...\n";






$qntd_label_positivo = 0;
$qntd_falso_negativo = 0;

for ($i=0;$i<count($testing->labels());$i++) {

	if ($testing->labels()[$i] == "S") {
		
		$qntd_label_positivo++;
		
		if ($predictions[$i] == "N") {
			
			$qntd_falso_negativo++;
			
		}
		
	}

}

echo "A precisão personalizada do modelo, com base nos elementos de teste, é de " . (($qntd_label_positivo - $qntd_falso_negativo)/$qntd_label_positivo*100) . "% (" . ($qntd_label_positivo - $qntd_falso_negativo) . "/" . $qntd_label_positivo . ")...\n";






/*
echo "Fazendo a previsão de um único elemento...\n";

$prediction = $estimator->predictSample([1,0,0,0,0,0,0,0,1,0,1,1,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1]);

echo "A previsão foi '" . $prediction . "'\n";
*/





echo "A matriz de confusão das previsões, com base nos elementos de teste, segue abaixo. Leia a matriz da seguinte forma: 'quando o modelo preveu não, X vezes deveria ter sido não e Y vezes deveria ter sido sim'. Lembrando que 'true negative' e 'true positive' é quando o modelo acerta e 'false negative' (o valor real é positivo mas o modelo preveu como negativo) e 'false positive' (o valor real é negativo mas o modelo preveu como positivo) é quando o modelo erra.\n";

$report = new ConfusionMatrix();
$result = $report->generate($predictions,$testing->labels());

var_dump($result);





echo "A importância de cada coluna é (nem todos os modelos permitem a execução desse recurso):\n";

$importances = $estimator->featureImportances();

asort($importances);

foreach ($importances as $chave1 => $valor1) {

	echo $coluna[$chave1] . " -> " . $valor1 . "\n";

}

/*
if (count($importances) != count($coluna)) {
	
	echo "Erro JE43. Resultado: " . count($importances) . "|" . (count($coluna) - 1);
	exit();
	
}
*/
/*
$temp32 = array();

for ($i=0;$i<count($importances);$i++) {

	$temp32[] = array($importances[$i],$coluna[$i]);
	
}

usort(

	$temp32,
	function ($a,$b){
		
		return ($a[0] < $b[0]?-1:1);

	}

);

for ($i=0;$i<count($temp32);$i++) {

	echo $temp32[$i][1] . " -> " . $temp32[$i][0] . "\n";
	
}
*/










echo "A importância das colunas principais, após RFE, é (nem todos os modelos permitem a execução desse recurso):\n";

$transformer = new RecursiveFeatureEliminator(5,2, new ClassificationTree(10));
$transformer->fit($dataset);
$importances = $transformer->importances();

asort($importances);

foreach ($importances as $chave1 => $valor1) {

	echo $coluna[$chave1] . " -> " . $valor1 . "\n";

}










$predictions1 = $estimator->predict($testing);

$qntd_diferencas_predictions_1_2 = array();

for ($i=0;$i<count($testing->samples()[0]);$i++) {
	
	$testing->transformColumn(

		$i,
		function ($value){
				
			if ($value === "N") {
			
				return "S";
			
			}
			else if ($value === "S") {
				
				return "N";
				
			}
			else{
				
				echo 'Erro JP90. Valor: "' . $value . '".';
				exit();
				//return $value;
				
			}
			
		}
		
	);	
	
	$predictions2 = $estimator->predict($testing);
	


	$qntd_diferencas_predictions_1_2[] = count(array_diff_assoc($predictions1,$predictions2));	
	
	
	
	$testing->transformColumn(

		$i,
		function ($value){
				
			if ($value === "N") {
			
				return "S";
			
			}
			else if ($value === "S") {
				
				return "N";
				
			}
			else{
				
				echo 'Erro JP90. Valor: "' . $value . '".';
				exit();
				//return $value;
				
			}
			
		}
		
	);



}



$temp32 = array();

for ($i=0;$i<count($qntd_diferencas_predictions_1_2);$i++) {

	$temp32[] = array($qntd_diferencas_predictions_1_2[$i],$coluna[$i]);
	
}

usort(

	$temp32,
	function ($a,$b){
		
		return ($a[0] < $b[0]?-1:1);

	}

);

echo "A importância de cada coluna, com base no indicador personalizado é:\n";

for ($i=0;$i<count($temp32);$i++) {

	echo $temp32[$i][1] . " -> " . $temp32[$i][0] . "\n";
	
}





?>
```

**Comment by andrewdalpino** [comment]
I understand @batata004 and I really hope you achieve your goals ... like I said there are a few issues with your training scripts. Why are you manually converting features to continuous when Random Forest handles categorical data natively for example? Some things stand out to me and it leads me to believe perhaps the best way to move forward is to take a step back for a moment and focus on the dataset (specifically feature engineering) and proper cross-validation.

I downloaded your dataset and played around with it for a bit. I couldn't come up with an intuitive explanation because I don't speak Portuguese but a couple things stand out. The first is that this is an imbalanced dataset (85% of the labels are no conversion, 15% are yes conversion) which would help to explain the bias toward predicting no conversion. Upsampling the positive samples via Random Forest results in the estimator basically predicting the prior probability of the class which, since there are two classes, after balancing it would be 50/50. This tells me that the features we are using are either not strongly correlated with the outcome/label or that the relationship between features and labels are quite complex and non-linear and cannot be captured by the learner (high bias typically indicates underfitting - see bias-variance tradeoff).

https://docs.rubixml.com/en/latest/choosing-an-estimator.html#bias-variance-tradeoff

I also experimented with [Naive Bayes](https://docs.rubixml.com/en/latest/classifiers/naive-bayes.html) and [AdaBoost](https://docs.rubixml.com/en/latest/classifiers/adaboost.html). Same issue with the bias, although Naive Bayes was able to allocate a tiny probability density to the yes label and in 2 cases and was able to make the correct positive prediction.

So it sounds to me that we are underfitting. To solve that we can start by adding more features. We're only using 6 features right now. If you don't have more feature data, you could try deriving some features from the data you have. I'd shoot for as many informative features as you can given the complexity of the problem (conversion prediction). If you see the bias starting to lift as you add more features, then you'll know to keep going.

https://docs.rubixml.com/en/latest/cross-validation.html#underfitting

In the meantime, I'll run some experiments on a few benchmark datasets to rule out any issues with the estimators themselves.

I hope this help!

**Comment by batata004** [comment]
My friend, you probably wont believe but I read most of the documentation 3 or 4 times, and some specific sessions I read at least 7 or 8 times! I tried really hard to follow the code and in my opinion, RUBIX ML has the easiest and simplest code possible to start running a ML training! I already took a look at other libraries and their code scaried me! Despite the "easy to use" of RUBIX ML, I am having a few problems.

First I will reply what you said in your last comment, ok? I dont understand why you said in my first code I am converting my dataset to continuous. I am not doing that! I am first importing the dataset -> hot encode -> convert to categorical. I need to convert to categorical because hotencode makes the data 0,1 which looks continuous to RandomForest and it throws an error. So I dont think I am doing anything wrong on the first code I posted. I am just doing the minimum necessary to run the code, right?

Also, my second code I know looks unnecessary long but it isnt! With that code I can save the model (persistence) and predict new data using the GET parameter "testar". I also built a custom score model to track specifically false positives (because I didnt like the results I was getting from `FBETA`) and I also created a custom model to select best features (because I didnt like the results I was getting from `RecursiveFeatureEliminator`).

I did lots of testing and unfortunatelly FBETA and RecursiveFeatureEliminator didnt look to solve my "questions". That's why that code is so long, I had to implement lots of things trying to understand better the training model since it looks like a "black box".

Let me say just one more thing: do you know what would be reaally nice to have on RUBIX ML? It shouldnt complain every time about the label/features not being continuous or categorical. If the model needs continuous features, then it should first check if the features have numbers on it (even if they are as strings) and try to work with those number. In case some string that cant be converted to number, THEN throw an error. ALSO if the model expects categorical features and if it finds numbers in the features, use the numbers as categories instead of complaning of the features/labels not being numerical - each different number is a category! 1 => "1", 2 => "2", 1.87 => "1.87"...

Do you know why I think this is important? Because every time I try to change the training model to another model, sometimes I need to go through all my features/labels and convert then to the expected values by the training models. Sometimes I just need to convert "1" to 1 or 1 to "1" just to make things work. That consumes lots of time. I think all training models, before start complaining about the datatype of the features/columns, it should FIRST try to work out with what it got. If it got numbers like 0,1,2,3... and needs categories, then just convert 0,1,2,3... to "0","1","2"... if the training model got strings like "0","1" and needs continuous values, then just try converting everything to numbers - in case something goes wrong THEN throws an error. This is something that really bothers me, besides the fact that I still cant make RUBIX ML run an example with stable results, **this issue that I am having of having really different results when running the same algorithm, in the same dataset, in the same order, without having anything random makes me too confused and I am losing trust in the library** (despite the fact that I am probably the one doing something wrong BUT even the simplest example I run, shows that the results are completely arbitrary).

**Comment by andrewdalpino** [comment]
> My friend, you probably wont believe but I read most of the documentation 3 or 4 times, and some specific sessions I read at least 7 or 8 times! I tried really hard to follow the code and in my opinion, RUBIX ML has the easiest and simplest code possible to start running a ML training! I already took a look at other libraries and their code scaried me! Despite the "easy to use" of RUBIX ML, I am having a few problems.

Great! What is it about the current docs that have been unclear such that the information we've tried to communicate is not making its way into your questions or code? Without knowing this, we are unable to make improvements that would have otherwise made your day alot easier.

> First I will reply what you said in your last comment, ok? I dont understand why you said in my first code I am converting my dataset to continuous. I am not doing that! I am first importing the dataset -> hot encode -> convert to categorical. I need to convert to categorical because hotencode makes the data 0,1 which looks continuous to RandomForest and it throws an error. So I dont think I am doing anything wrong on the first code I posted. I am just doing the minimum necessary to run the code, right?

One Hot Encoder is for converting categorical features to continuous. See the docs here https://docs.rubixml.com/en/latest/transformers/one-hot-encoder.html. It looks like you are converting categorical features to continuous and then back to categorical again. I've also mentioned this in one of your previous questions [here](https://github.com/RubixML/RubixML/issues/97#issuecomment-664634472).

> Let me say just one more thing: do you know what would be reaally nice to have on RUBIX ML? It shouldnt complain every time about the label/features not being continuous or categorical. If the model needs continuous features, then it should first check if the features have numbers on it (even if they are as strings) and try to work with those number. In case some string that cant be converted to number, THEN throw an error. ALSO if the model expects categorical features and if it finds numbers in the features, use the numbers as categories instead of complaning of the features/labels not being numerical - each different number is a category! 1 => "1", 2 => "2", 1.87 => "1.87"...

What you are referring to here is AutoML which is an area we continuously work on. This would not be handled by the learner however, it would most likely be implemented as a meta-estimator. I'd suggest trying a higher-level cloud platform if you are expecting everything to be done for you ex. https://cloud.google.com/automl however your results will still be bad if you have a bad dataset. For now, we have a few AutoML features you can use and we have more on the way such as [Recursive Feature Eliminator](https://github.com/RubixML/Extras/blob/master/docs/transformers/recursive-feature-eliminator.md) in the Extras package.

https://docs.rubixml.com/en/latest/automl.html

> Do you know why I think this is important? Because every time I try to change the training model to another model, sometimes I need to go through all my features/labels and convert then to the expected values by the training models. Sometimes I just need to convert "1" to 1 or 1 to "1" just to make things work. That consumes lots of time. I think all training models, before start complaining about the datatype of the features/columns, it should FIRST try to work out with what it got. If it got numbers like 0,1,2,3... and needs categories, then just convert 0,1,2,3... to "0","1","2"... if the training model got strings like "0","1" and needs continuous values, then just try converting everything to numbers - in case something goes wrong THEN throws an error. This is something that really bothers me, besides the fact that I still cant make RUBIX ML run an example with stable results, this issue that I am having of having really different results when running the same algorithm, in the same dataset, in the same order, without having anything random makes me too confused and I am losing trust in the library (despite the fact that I am probably the one doing something wrong BUT even the simplest example I run, shows that the results are completely arbitrary).

Welcome to the world of data science and data engineering! Have you tried one of our example projects with a properly clean/engineered dataset? What do you mean by stable results?

**Comment by andrewdalpino** [comment]
I ran a test on Random Forest with the [mushroom datatset](https://archive.ics.uci.edu/ml/datasets/Mushroom) which is a two-class (poisonous/non-poisonous) dataset similar to yours in that it contains alot of categorical features however it has 22 of them instead of 6. I used the settings of your hyper-parameters as well as an optimized set and in each case I was able to achieve 'stable' predictions. I've attached screencaps of the confusion matrices from 3 separate training and validation runs. Despite Random Forest being a non-deterministic algorithm, you can see that it still converges to a good solution because it found patterns in the dataset. I did not use one-hot encoding either because that is not necessary with Random Forest.

So we know that with a cleaned and properly engineered dataset we can achieve stability so this points me back to your dataset and adding more features.

<img width="509" alt="Annotation 2020-07-29 002910" src="https://user-images.githubusercontent.com/18690561/88760462-b1379000-d132-11ea-9dda-26110c2aa945.png">

<img width="210" alt="Annotation 2020-07-29 002243" src="https://user-images.githubusercontent.com/18690561/88760471-b563ad80-d132-11ea-9453-9c4ead9c5393.png">

<img width="186" alt="Annotation 2020-07-29 002320" src="https://user-images.githubusercontent.com/18690561/88760479-b98fcb00-d132-11ea-88b6-f483e81148ef.png">

<img width="189" alt="Annotation 2020-07-29 002359" src="https://user-images.githubusercontent.com/18690561/88760488-bbf22500-d132-11ea-89a5-a3d6dd22dc65.png">

In addition to the confusion matrices you said that your feature importances varied alot as well but I was able to achieve stable results with the same settings as above with the mushroom dataset. You'll notice that feature column 4 is always the first, 10, 19, and 20 are always in the top 5, and column 5, 15, and 16 are always one of the last few in the sorted list of feature importances.

<img width="221" alt="Annotation 2020-07-29 004042" src="https://user-images.githubusercontent.com/18690561/88761336-8ea67680-d134-11ea-9c50-c3f52aa24c1b.png">

<img width="223" alt="Annotation 2020-07-29 004119" src="https://user-images.githubusercontent.com/18690561/88761342-9239fd80-d134-11ea-8919-20c39d7f6078.png">

<img width="218" alt="Annotation 2020-07-29 004158" src="https://user-images.githubusercontent.com/18690561/88761360-96feb180-d134-11ea-946c-6169e2f14184.png">

**Comment by batata004** [comment]
> One Hot Encoder is for converting categorical features to continuous. See the docs here https://docs.rubixml.com/en/latest/transformers/one-hot-encoder.html. It looks like you are converting categorical features to continuous and then back to categorical again. I've also mentioned this in one of your previous questions here.

I understand what you said, but there is also another great use of One Hot Encoder! Assume I have a specific feature with 4 categorical values "A","B","C","D". I dont like the fact that a column has 4 different values because, in my future analysis trying to understand the model, it becomes very dificult to know which feature value was more important. The nice thing about One Hot Encoder is that if I use it on a feature with multiple values, it creates a new feature for every distinct value! Then, if `featureImportances` produced a reliable output, it could help me understand NOT ONLY the features/columns which are most important, it would help me to understand which feature's values are more important too since every single distinct value of every single feature, was break down to separated features!

That's why I use One Hot Encoder to break down all distinct values, so I am able, later, to understand every single feature's value importance. 

Because of that I need to convert all categorical features to continuous (with One Hot Encoder) and then, later, convert it back to categorical. IF RUBIX ML could assume 0's and 1's as categorical values, I wouldnt need that last step to convert 1 to "S" (yes) and 0 to "N" (no). Unfortunatelly, Rubix ML is very strict about this, while being very loose regarding other things.

> What you are referring to here is AutoML which is an area we continuously work on. This would not be handled by the learner however, it would most likely be implemented as a meta-estimator. I'd suggest trying a higher-level cloud platform if you are expecting everything to be done for you ex. https://cloud.google.com/automl however your results will still be bad if you have a bad dataset. For now, we have a few AutoML features you can use and we have more on the way such as Recursive Feature Eliminator in the Extras package.

I think I didnt make myself clear! When you replied this to me, I was complaning about the fact that RUBIX ML is too strict about the data type of the dataset when using specific models. Some models cant simply accept the fact the 0 could be "0" and 1 could be "1" and then work. No, some models require me explicitly to convert 0 to "N" and 1 to "S". I dont know why the models are too strict about it, I think RUBIX ML models should accept any type value in the input dataset, if the dataset is continuous and the model requires categorial, then the model should automatically convert the continuous values to any category value. And if the dataset is categorical and the model requires continuous, then the model should try to convert all categorical values to numerical, if in case any string cant be converted to a number THEN, and only the, rubix ml should throw an error.

> Welcome to the world of data science and data engineering! Have you tried one of our example projects with a properly clean/engineered dataset?

Yeap, I tried almost all of your examples! But things get way easier when you already have a dataset that was carefully selected to produce good results. It's pretty easy generating a pretty nice dataset, with 99% perfect correlation (R^2) between features and labels. And of course, with a pretty nice dataset, any ML model, no matter how bad programmed it was, it will eventually display a good accuracy. Now, trying a real world dataset is the real proof if a ML library is really doing a great job.

>  What do you mean by stable results?

When I mean stable results I dont want the results of the ML model to be good (high accuracy) or bad (low accuracy). I just want the results to be the same, no matter how many times I run the same model, using the same dataset, the same hyperparameters... it should always output a very similar result. It's makes me pretty hard do believe in the results of a ML library that running the same code, the same dataset, same confgurations... outputs completely random results. The biggest problem so far with Rubix ML is making me believe that the results it produces are indeed worth any credit. Sorry my friend, but despite how bad or good my dataset is, the results should be consistently bad or good. The accuracy of a model cant be close to 70% and a second later close to 3% if it was trained using the same model, same values, same dataset... same everything! Of course, as I said other times, I know most ML models introduce bias, but they are small and should not result in a 3% accuracy at one second and 70% accuracy in another second. No matter how bad or good the dataset is!

And I know RandomForest has `random` in its name, but my friend executed the same model in Tensorflow and he got bad results (not good accuracy), however every time he execuetd the code, he told me he would get veeery similar (if not exact) results. Running the same model using RubixML is a completely different story, the results, the importance feature, the confusion matrix... changes a lot everytime I execute the same exact code.

>  Despite Random Forest being a non-deterministic algorithm

As I said above, I understand that RandomForest adds lots of bias (I imagine it introduces more bias than any other model I tried because the name itself of this model implies `random`) however those bias should not produce literally complete different results. Also, even if RandomForest uses lots of bias, it should generate the same seed everytime it runs. If you think this is a bad ideia, maybe you could allow the user to set a string seed (as a hyperparameter) everytime the user trains a model. With the same exact seed, all future "random" biases or "random" decisions made by the model, would be the same, no matter how many times I run the model. I dont know if I am making myself clear into this, but enabling the user to set a seed to the random generator, it would make the model completely 100% perfect exact! And sometimes it's very good to have 100% perfect match between runs beucase I may be trying to evaluate how good 2 different datasets are. If I know for a fact that my model runs 100% exactly the same 100% of the time, I can comparete 2 datasets and make real decisions regarding how good one is and how bad the other is. I could even manually set 3 different seeds to check the results of each dataset. 

What I cant believe as acceptable, is running the same dataset, in the same model, same everything... and then getting completely random results. It makes pretty hard to 1) trus the library 2) trust the model and 3) trust if the dataset I have can be comparable to another dataset.

> In addition to the confusion matrices you said that your feature importances varied alot as well but I was able to achieve stable results with the same settings as above with the mushroom dataset. You'll notice that feature column 4 is always the first, 10, 19, and 20 are always in the top 5, and column 5, 15, and 16 are always one of the last few in the sorted list of feature importances.

I really appreciate your kindness taking lots of your time to replicate my code on your end. I really appreciate your help my friend! But what you did, just prooves one thing: your dataset is better than mine. You used a dataset that surelly is 100% optimized to produce good results to showcase ML. I am very sure 99% of the datasets around on the internet, that many libraries use, are 90% "optimized" (hand selected) to work good with most models. You surelly wouldnt showcase a ML library, or provide examples to a ML library, with a dataset that is subject to lots of noise (which MOST of the real case dataset have, LOTS of noise!). 

So, with all respect, the stable results you proved you got running my code with your dataset just shows one thing: your dataset is way more optimized than mine. It does not prove that the model, or the library, was well programmed (please, I am not offending anyone here, I am 100% grateful for this community help), it just proves that your dataset is pretty "curated" (biased to produce good results) and not a real use case because real datasets, most of the time, have LOTS of noise. If there wasnt noise, it would be pretty easy to use any SOLVE method of excel to make predictions on any data. It's exactly the good noise dealing that makes ML very good (there exists many other reasons but I think the major advtange of ML is having it to work well with noises, much better than any SOLVE/linear/logaritmic regression embeded in excel).
